File: Week 03/id_377/122.best-time-to-buy-and-sell-stock-ii.js
/*
 * @lc app=leetcode id=122 lang=golang
 */

// @lc code=start
// ------------------------- 解法 1 start --------------------------
// 最好记  👍 👍 👍 👍 👍
// 最优解  😘 😘 😘 😘 😘
var maxProfit = function(prices) {
    let diff = 0
    if (prices.length > 0) { // []
        prices.reduce((acc, next) => {
            if (next > acc) {
                diff += next - acc
            }
            return next
        })
    }
    return diff
};
// ------------------------- 解法 2 start --------------------------
// 最好记  👍 👍 👍 👍
// 最优解  😘 😘 😘 😘
var maxProfit = function(prices) {
    let max = 0;
    for (let i = 1; i < prices.length; i++) {
        if (prices[i] > prices[i - 1]) max += prices[i] - prices[i - 1];
    }
    return max;
}

// @lc code=end

File: Week 03/id_377/122.best-time-to-buy-and-sell-stock-ii.go
/*
 * @lc app=leetcode id=122 lang=golang
 */

// @lc code=start
// ------------------------- 解法 1 start --------------------------
// 最好记  👍 👍 👍 👍 👍
// 最优解  😘 😘 😘 😘 😘
func maxProfit(prices []int) int {
	if len(prices) == 0 {
		return 0
	}
	profit := 0
	min := prices[0]

	for _, price := range prices {
		if price > min {
			profit += price - min
		}
		min = price
	}
	return profit
}

// @lc code=end

